### PR TITLE
Issue 619 changes log

### PR DESCRIPTION
**Pull request by artemhoriuk1994**
Pull Request Template
 

### Type of Pull Request *
- [ ] CHANGE (fix or feature that would cause existing functionality to not work as expected)
- [ ] FEATURE (non-breaking change which adds functionality)
- [ ] BUGFIX (non-breaking change which fixes an issue)
- [ x ] ENHANCEMENT  (non-breaking change which improves existing functionality)
- [ ] NONE (if none of the other choices apply. For example, tooling, build system, CI, docs, etc.)

#
 
### Description *
Did everything according to story #619. The option to export data to a .csv file remained unfulfilled.

image

### CODE REVIEW

File: src/old/lib/utilities/API/api.ts [hunk 1]
```diff
@@ -306,6 +308,16 @@ export const getPlatformInformation = async (
   return instance.get(`/platform-information/${id}`);
 };
 
+export const fetchMaterialsChange = async (
```
Comment (skyboyer) [inline src/old/lib/utilities/API/api.ts:311]: I propose to name on `fetchChangeLog` or in some other way but to make `changeLog` to be part of the name

File: src/views/Profile/AdminTable/AdminDatePicker.tsx [hunk 2]
```diff
@@ -65,7 +65,7 @@ export const AdminDatePicker: React.FC<IMaterialsDate> = ({
               onChange={(value) =>
                 setChanges({
                   option: 'start',
-                  date: value && value.toISOString(),
+                  date: value && value.toISOString().substring(0, 10),
```
Comment (skyboyer) [inline src/views/Profile/AdminTable/AdminDatePicker.tsx:68]: I propose to introduce helper in /src/utilities/formatDate.ts for date formating

1. shorter code
2. can cover with tests(also for edge cases)
3. consistent formatting across different occurences

do you agree?

File: src/views/postUpdation/AllPostTypesUpdation.tsx [hunk 3]
```diff
@@ -444,6 +479,31 @@ export const AllPostTypesUpdation: React.FC<IAllPostTypesUpdation> = ({
...
+        <AuthorListDropdown
+          pageSizes={changesPerPage}
+          setChanges={handleNotesToShowChange}
+          selected={
+            size === 0 ? changesPerPage[changesPerPage.length - 1] : size
+          }
+        />
+        <List>
+          {changes.length > 0 ? (
+            changes.map((item) => {
```
Comment (skyboyer) [inline src/views/postUpdation/AllPostTypesUpdation.tsx:494]: another approach is to render `changes.map` always:

```
<List>
  {changes.map(item => .....)}
  {changes.length === 0 && <>No Changes</>}
```

Definitely no difference in behaviour. But to me this variant is easier to follow, than ternary `? :`

File: src/views/postUpdation/AllPostTypesUpdation.tsx [hunk 4]
```diff
@@ -444,6 +479,31 @@ export const AllPostTypesUpdation: React.FC<IAllPostTypesUpdation> = ({
...
         post={post}
       />
+      <BorderBottom />
+      <h1>Changes List</h1>
+      <Box mt={3}>
+        <AuthorListDropdown
+          pageSizes={changesPerPage}
+          setChanges={handleNotesToShowChange}
+          selected={
+            size === 0 ? changesPerPage[changesPerPage.length - 1] : size
```
Comment (skyboyer) [inline src/views/postUpdation/AllPostTypesUpdation.tsx:489]: let's go with `size || changesPerPage[changesPerPage.length - 1]`

not just shorter(code size should never be an argument per se) but also covers edge case with `undefined`(not sure whether it's allowed for this type but anyway)

File: src/views/postUpdation/AllPostTypesUpdation.tsx [hunk 5]
```diff
@@ -444,6 +479,31 @@ export const AllPostTypesUpdation: React.FC<IAllPostTypesUpdation> = ({
...
+            changes.map((item) => {
+              return (
+                <ListItem key={item.id}>
+                  {item.title} {item.changes}:{' '}
+                  {new Date(item.dateOfChange).toUTCString()}
+                </ListItem>
+              );
+            })
+          ) : (
+            <>No Chanages</>
```
Comment (skyboyer) [inline src/views/postUpdation/AllPostTypesUpdation.tsx:503]: may we want i18n here?

File: src/views/postUpdation/AllPostTypesUpdation.tsx [hunk 6]
```diff
@@ -444,6 +479,31 @@ export const AllPostTypesUpdation: React.FC<IAllPostTypesUpdation> = ({
         disabled={isDisabled}
         post={post}
       />
+      <BorderBottom />
+      <h1>Changes List</h1>
```
Comment (skyboyer) [inline src/views/postUpdation/AllPostTypesUpdation.tsx:483]: also, may we want to keep it in i18n strings?

File: src/models/changeLog/asyncAction.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import { fetchMaterialsChange } from 'old/lib/utilities/API/api';
+import { IChangeLog } from './types';
+
+export const getChangeLog = createAsyncThunk(
```
Comment (skyboyer) [inline src/models/changeLog/asyncAction.ts:5]: but where this action is used?

File: src/views/Profile/AdminTable/AdminTableBody.tsx [hunk 8]
```diff
@@ -68,9 +69,11 @@ const AdminTableBody: React.FC<IAdminTableBodyProps> = ({ isAdmin }) => {
         <TableCell>{id}</TableCell>
         <TableCell className={classes.titleCol}>
           <Chip label={typeName} className={classes[chipClass] as string} />
-          <Typography variant="h6" component="p" className={classes.title}>
-            {title}
-          </Typography>
+          <Link to={`/edit-post?id=${id}`} className={classes.title}>
```
Comment (skyboyer) [inline src/views/Profile/AdminTable/AdminTableBody.tsx:72]: we cannot use this here, but just as knowledge sharing: there is `generatePath` helper in `react-router-dom`

so if `/edit-post/:id` was a named route(which it is NOT and `id` parameter in query part instead of string - but let's imagine) then we could:

```
import { EDIT_POST_ROUTE } from '../...../routes.js';

  <Link to={generatePath(EDIT_POST_ROUTE, { postId: id })}

```

but here we cannot use that trick, so just saying

File: src/models/changeLog/selectors.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import { RootStateType } from 'models/rootReducer';
+
+export const selectSize = (state: RootStateType) => {
+  return state.changesLog;
```
[the remark this answers is not in the document]
Comment (artemhoriuk1994) [reply src/models/changeLog/selectors.ts:4]: so it should be state.changeLog.size
First, I got the size object through destructuring in the component
So im chnge that for more better readability

File: src/views/postUpdation/AllPostTypesUpdation.tsx (hunk 4 above)
Comment (artemhoriuk1994) [reply src/views/postUpdation/AllPostTypesUpdation.tsx:489]: yeap im agree with you that this one way look more better

File: src/models/changeLog/asyncAction.ts (hunk 7 above)
Comment (artemhoriuk1994) [reply src/models/changeLog/asyncAction.ts:5]: this acition im used in reducet like extraReducer 
```suggestion
const changeLog = createSlice({
  name: 'changesLog',
  initialState,
  reducers: {
    setChangesSize: (state, action: PayloadAction<number>) => {
      const setChangeObj = state;
      setChangeObj.size = action.payload;
    },
  },
  extraReducers: {
    getAsyncActionsReducer(getChangeLog as any),
  },
});
```

File: src/views/Profile/AdminTable/AdminDatePicker.tsx (hunk 2 above)
Comment (artemhoriuk1994) [reply src/views/Profile/AdminTable/AdminDatePicker.tsx:68]: Create function helper for this date request

File: src/old/lib/components/Chips/ChipsList.tsx [hunk 10]
```diff
@@ -9,11 +9,7 @@ export interface IChipsListProps {
   max?: boolean;
 }
 
-const ChipsList: React.FC<IChipsListProps> = ({
-  checkedNames,
-  isLabelItem,
-  max,
-}) => {
+const ChipsList: React.FC<IChipsListProps> = ({ checkedNames }) => {
```
Comment (skyboyer) [inline src/old/lib/components/Chips/ChipsList.tsx:12]: since we don't use `max` and `isLabelItem` anymore, may it make sense to remove them from `IChipsListProps` as well?

File: src/old/lib/utilities/API/types.ts [hunk 11]
```diff
@@ -277,3 +278,36 @@ export type UpdatePlatformInformationRequestType = {
   title?: string;
   text?: string;
 };
+
+export type ChangeLogType<T = any> = {
```
Comment (skyboyer) [inline src/old/lib/utilities/API/types.ts:282]: seems `T` is not used in this generic, so do we need it at all here?

File: src/old/modules/experts/components/Accordion.tsx [hunk 12]
```diff
@@ -1,5 +1,4 @@
 import React, { useState } from 'react';
-import { NavLink } from 'react-router-dom';
```
Comment (skyboyer) [inline src/old/modules/experts/components/Accordion.tsx:2]: more cleanup! I like that!

File: src/styles/theme/index.ts [hunk 13]
```diff
@@ -10,7 +10,7 @@ export const SCREEN_BREAKPOINTS = {
   xl: 1920,
 };
 
-export const MAIN_THEME = createMuiTheme({
+export const MAIN_THEME = createTheme({
```
Comment (skyboyer) [inline src/styles/theme/index.ts:13]: yeeeey! no more deprecated function!

File: src/utilities/formatDate.ts [hunk 14]
```diff
@@ -42,3 +42,7 @@ export const displayShortDate = (
   }
   return `${dateValue}`;
 };
+
+export const requestDate = (value: Date): string => {
+  return value && value.toISOString().substring(0, 10);
```
Comment (skyboyer) [inline src/utilities/formatDate.ts:47]: since `value` is declared as a `Date` there is no need to do `value &&`. we cannot pass `undefined` or `null`. right?

File: src/views/Profile/AuthorsList/AuthorListDropdown.tsx [hunk 15]
```diff
@@ -3,9 +3,9 @@ import { FormControl, Select, InputBase, MenuItem } from '@material-ui/core';
 import { useStyles } from './styles/AuthorListDropdown.styles';
 
 interface IAuthorListDropdownProps {
-  pageSizes: number[];
-  selected: number;
-  setChanges: (payload: any) => void;
+  pageSizes: Array<number | string>;
```
Comment (skyboyer) [inline src/views/Profile/AuthorsList/AuthorListDropdown.tsx:6]: I believe we should strict: `pageSizes: Array<number | 'All'>`; similarly to `selected` property.

File: src/views/Profile/AdminTable/AdminTableBody.tsx [hunk 16]
```diff
@@ -68,9 +69,17 @@ const AdminTableBody: React.FC<IAdminTableBodyProps> = ({ isAdmin }) => {
         <TableCell>{id}</TableCell>
         <TableCell className={classes.titleCol}>
           <Chip label={typeName} className={classes[chipClass] as string} />
-          <Typography variant="h6" component="p" className={classes.title}>
-            {title}
-          </Typography>
+          <Link
+            to={{
+              pathname: '/edit-post',
```
Comment (skyboyer) [inline src/views/Profile/AdminTable/AdminTableBody.tsx:74]: cool! I like object-style target for `<Link>`

File: src/views/postUpdation/AllPostTypesUpdation.tsx [hunk 17]
```diff
@@ -444,6 +478,34 @@ export const AllPostTypesUpdation: React.FC<IAllPostTypesUpdation> = ({
...
       />
+      <BorderBottom />
+      <Typography component="h2" variant="h2">
+        {t(langTokens.admin.changesList)}
+      </Typography>
+      <Box mt={3}>
+        <AuthorListDropdown
+          pageSizes={changesPerPage}
+          setChanges={handleNotesToShowChange}
+          selected={size || changesPerPage[changesPerPage.length - 1]}
```
Comment (skyboyer) [inline src/views/postUpdation/AllPostTypesUpdation.tsx:489]: if we update `IAuthorListDropdownProps` to allow only `All` among other strings, we could also hard-code it here:

```
selected={size || 'All'}
```
and TS will keep us safe

**Review by skyboyer — COMMENTED** [review]
looks good, have few comments

File: src/old/lib/components/Chips/ChipsList.tsx (hunk 10 above)
Comment (artemhoriuk1994) [reply src/old/lib/components/Chips/ChipsList.tsx:12]: I think its  better to move `IChipsListProps` to a new file, cause another components use this interfece

File: src/utilities/formatDate.ts (hunk 14 above)
Comment (artemhoriuk1994) [reply src/utilities/formatDate.ts:47]: yeap thats my fault 
im tottaly agree with u

File: src/views/Profile/AuthorsList/AuthorListDropdown.tsx (hunk 15 above)
Comment (artemhoriuk1994) [reply src/views/Profile/AuthorsList/AuthorListDropdown.tsx:6]: We have problem like this
image

Comment (skyboyer) [reply src/views/Profile/AuthorsList/AuthorListDropdown.tsx:6]: give me some time, I'll play with it

File: src/views/Profile/AuthorsList/AuthorListDropdown.tsx [hunk 18]
```diff
@@ -2,10 +2,12 @@ import React from 'react';
...
 import { useStyles } from './styles/AuthorListDropdown.styles';
 
+export type PageSizeType = Array<number | 'All'>;
+
 interface IAuthorListDropdownProps {
-  pageSizes: number[];
-  selected: number;
-  setChanges: (payload: any) => void;
+  pageSizes: PageSizeType;
+  selected: number | string | undefined;
```
Comment (skyboyer) [inline src/views/Profile/AuthorsList/AuthorListDropdown.tsx:9]: but what if we had `type PageSizeType = number | 'All'`? Then we could type `selected: PageSizeType | undefined`

or does it create any issues in other code?

File: src/views/Profile/AuthorsList/AuthorListDropdown.tsx [hunk 19]
```diff
@@ -2,10 +2,12 @@ import React from 'react';
 import { FormControl, Select, InputBase, MenuItem } from '@material-ui/core';
 import { useStyles } from './styles/AuthorListDropdown.styles';
 
+export type PageSizeType = Array<number | 'All'>;
```
Comment (skyboyer) [inline src/views/Profile/AuthorsList/AuthorListDropdown.tsx:5]: as for naming, may it use plural like `PageSizesType` for an array?

**Review by skyboyer — APPROVED** [review]
looks good to me, just 2 minor comments
